base/special/exp2.jl

```diff
@@ -0,0 +1,445 @@
+#     Thus we have exp2f(x) = 2^k * exp2(y).
+#
+#   Reduce y:
+#     y = i/TABLESIZE + z for integer i near y * TABLESIZE.
+#     Thus we have exp2(y) = exp2(i/TABLESIZE) * exp2(z),
+#     with |z| <= 2^-(TABLESIZE+1).
+#
+#   We compute exp2(i/TABLESIZE) via table lookup and exp2(z) via a
+#   degree-4 minimax polynomial with maximum error under 1.4 * 2^-33.
+#   Using double precision for everything except the reduction makes
+#   roundoff error insignificant and simplifies the scaling step.
+#
+#   This method is due to Tang, but I do not use his suggested parameters:
+#
+#	Tang, P.  Table-driven Implementation of the Exponential Function
+#	in IEEE Floating-Point Arithmetic.  TOMS 15(2), 144-157 (1989).
+
+exp2_tablebits(::Type{Float64}) = UInt32(8)
+exp2_tablebits(::Type{Float32}) = UInt32(4)
+exp2_tablesize(T) = UInt32(1) << exp2_tablebits(T)
+
+exp2_overflow(::Type{Float64}) = 1024.0
+exp2_overflow(::Type{Float32}) = 128f0
+exp2_underflow(::Type{Float64}) = -1075.0
+exp2_underflow(::Type{Float32}) = -150f0
+exp2_small(::Type{Float64}) = 5.551115123125783e-17 # 2.0^-54
+exp2_small(::Type{Float32}) = 5.9604645f-8 # 2f0^-24
+@inline exp2_word(x::Float64) = reinterpret(UInt64, x) % UInt32 # low word
+@inline exp2_word(x::Float32) = reinterpret(UInt32, x) # word
+
+@inline exp2_float_from_words(highword, lowword) = reinterpret(Float64, (UInt64(highword)<<32)|lowword)
+@inline exp2_float_from_words(highword, ::Nothing ) = reinterpret(Float32, highword)
+
+
+exp2_redux(T::Type{Float64}) = 6.755399441055744e15 / exp2_tablesize(T) # 3*2.0^51 / exp2_tablesize(T)
+exp2_redux(T::Type{Float32}) = 1.2582912f7 / exp2_tablesize(T) # 3*2f0^22 / exp2_tablesize(T)
+
+function exp2(x::T) where T<:Union{Float32, Float64}
+    absx = abs(x)
+
+    # Filter out exceptional cases
+    isinf(x) && return ifelse(signbit(x), T(0), x)
+    isnan(x) && return x
+    if absx >= exp2_overflow(T)
+        x >= exp2_overflow(T) && return T(Inf) # overflow  without signal
+        x <= exp2_underflow(T) && return zero(T) # underflow without signal
+    elseif absx < exp2_small(T)
+        return one(T) + x
+    end
+
+    # Reduce x, computing z, i0, and k.
+    t = x + exp2_redux(T)
+    i0 = exp2_word(t)
+    i0 += exp2_tablesize(T)>>1
+    k = ((i0 >> exp2_tablebits(T)) << 20) % Int32
+    i0 &= exp2_tablesize(T) - UInt32(1)
+    if T == Float64
+        i0 = i0 << 1
+    end
+    i0 += UInt32(1) # reference index is 1 not 0
+    t -= exp2_redux(T)
+    z = x - t
+
+    # Compute r = exp2(y) and scale by 2^(k>>20)
+    # based on x being Float32 or Float64
+    exp2_kernel(k, i0, z, x)
+end
+
+@inline function exp2_kernel(k, i0, z, x::Float64)
+    # Compute r = exp2(y) = exp2t[i0] * p(z - eps[i]).
+    t = exp2_combined_table[i0]	# exp2t[i0]
+    z -= exp2_combined_table[i0 + 1] # eps[i0]
+    if k >= -(1021 << 20)
+        twopk = exp2_float_from_words(0x3ff00000 + k, 0)
+    else
+        twopkp1000 = exp2_float_from_words(0x3ff00000 + k + (1000 << 20), 0)
+    end
```

Comment:
I don't understand why this `k >= -(1021 << 20)` check is here.  Couldn't you merge this with the `k >= -(1021 << 20)` check below, since `twopk` etc. is not needed until after the `@horner` call?

Reply:
(Indeed, you don't really need variables `twopk` and `twopkp1000` at all, since you only use them once.  Just merge these expressions into `r * twopk` etc. below.)

Reply:
good points